File: ios/Honoured/Auth/Core/AuthSupport.swift
import CryptoKit
import Foundation

/// Only one system sign-in UI at a time. Apple and Google share this lock, and
/// a Google cleanup holds it too, so a sign-out still running cannot overlap a
/// new sign-in. Main thread only.
final class ProviderPresentationGate {
    enum Holder: Equatable {
        case apple
        case google
        case googleCleanup
    }

    private(set) var holder: Holder?
    private var token = 0

    /// Returns a token for `release`, or nil when another holder is active.
    func acquire(_ holder: Holder) -> Int? {
        guard self.holder == nil else { return nil }
        self.holder = holder
        token += 1
        return token
    }

    /// Releases only the holder that owns `token`; a stale release is ignored.
    func release(_ token: Int) {
        guard token == self.token else { return }
        holder = nil
    }
}

enum AuthNonce {
    /// 32 random bytes as 64 hex characters, and the SHA-256 hex digest of that
    /// string. The provider sees the digest; Supabase gets the raw value.
    static func make() -> (raw: String, hashed: String) {
        var bytes = [UInt8](repeating: 0, count: 32)
        if SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) != errSecSuccess {
            var generator = SystemRandomNumberGenerator()
            bytes = (0..<32).map { _ in UInt8.random(in: .min ... .max, using: &generator) }
        }
        let raw = bytes.map { String(format: "%02x", $0) }.joined()
        return (raw, sha256Hex(raw))
    }

    static func sha256Hex(_ input: String) -> String {
        SHA256.hash(data: Data(input.utf8)).map { String(format: "%02x", $0) }.joined()
    }
}

/// Public Google client configuration. None of these values is a secret, but a
/// malformed or missing one must turn the feature off instead of letting the
/// SDK raise an exception at presentation time.
struct GoogleClientConfig: Equatable {
    let clientID: String
    let serverClientID: String
    let callbackScheme: String

    private static let suffix = ".apps.googleusercontent.com"
    private static let schemePrefix = "com.googleusercontent.apps."

    /// - Parameter registeredSchemes: every `CFBundleURLSchemes` entry of the
    ///   app. The SDK requires the reversed iOS client ID among them.
    static func validate(
        clientID rawClientID: String,
        serverClientID rawServerClientID: String,
        registeredSchemes: [String]
    ) -> GoogleClientConfig? {
        let clientID = rawClientID.trimmingCharacters(in: .whitespacesAndNewlines)
        let serverClientID = rawServerClientID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard isClientID(clientID), isClientID(serverClientID), clientID != serverClientID else { return nil }
        let scheme = schemePrefix + String(clientID.dropLast(suffix.count))
        guard registeredSchemes.contains(where: { $0.caseInsensitiveCompare(scheme) == .orderedSame }) else { return nil }
        return GoogleClientConfig(clientID: clientID, serverClientID: serverClientID, callbackScheme: scheme)
    }

    private static func isClientID(_ value: String) -> Bool {
        guard value.hasSuffix(suffix), value.count > suffix.count else { return false }
        let prefix = value.dropLast(suffix.count)
        return prefix.allSatisfy { $0.isLetter || $0.isNumber || $0 == "-" }
    }
}

/// Where an incoming URL goes. Only the configured Google callback scheme
/// reaches the Google SDK; everything else keeps its existing route.
enum AppURLRoute: Equatable {
    case googleCallback
    case app

    static func classify(_ url: URL, googleCallbackScheme: String?) -> AppURLRoute {
        if let googleCallbackScheme,
           let scheme = url.scheme,
           scheme.caseInsensitiveCompare(googleCallbackScheme) == .orderedSame {
            return .googleCallback
        }
        return .app
    }
}

/// The exact origin (scheme, host, port) the auth bridge trusts, derived from
/// the configured web app URL. No wildcard and no suffix matching.
struct TrustedWebOrigin: Equatable {
    let scheme: String
    let host: String
    let port: Int

    init?(url: URL) {
        guard let scheme = url.scheme?.lowercased(), scheme == "https",
              let host = url.host?.lowercased(), !host.isEmpty else { return nil }
        self.scheme = scheme
        self.host = host
        self.port = url.port ?? 443
    }

    func matches(scheme: String?, host: String?, port: Int?) -> Bool {
        guard let scheme = scheme?.lowercased(), let host = host?.lowercased() else { return false }
        // WebKit reports 0 for the default port of the scheme.
        let resolvedPort = (port == nil || port == 0) ? (scheme == "https" ? 443 : -1) : port!
        return scheme == self.scheme && host == self.host && resolvedPort == self.port
    }

    func matches(url: URL?) -> Bool {
        guard let url else { return false }
        return matches(scheme: url.scheme, host: url.host, port: url.port)
    }
}

/// Runs async operations strictly one after another, in submission order.
/// Used for auth session writes and billing identity changes, so a slow older
/// operation can never finish after a newer one. Submission order is the
/// order of `enqueue` calls; submit from one thread to make that meaningful.
final class SerialAsyncQueue {
    private let lock = NSLock()
    private var tail: Task<Void, Never>?

    func enqueue(_ operation: @escaping () async -> Void) {
        lock.lock()
        defer { lock.unlock() }
        let previous = tail
        tail = Task {
            await previous?.value
            await operation()
        }
    }

    /// Enqueues and waits for this operation's own result.
    func run<T>(_ operation: @escaping () async -> T) async -> T {
        await withCheckedContinuation { continuation in
            enqueue {
                continuation.resume(returning: await operation())
            }
        }
    }
}

File: android/app/src/main/java/com/honoured/app/auth/AuthSupport.kt
package com.honoured.app.auth

import java.net.URI
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.ArrayDeque
import java.util.concurrent.atomic.AtomicBoolean

/**
 * One Google presentation or cleanup at a time; a cleanup still running blocks
 * a new sign-in. Main thread only.
 */
class ProviderPresentationGate {
    enum class Holder { GOOGLE, GOOGLE_CLEANUP }

    var holder: Holder? = null
        private set
    private var token = 0

    fun acquire(holder: Holder): Int? {
        if (this.holder != null) return null
        this.holder = holder
        token += 1
        return token
    }

    /** Releases only the holder that owns [token]; a stale release is ignored. */
    fun release(token: Int) {
        if (token == this.token) holder = null
    }
}

object AuthNonce {
    private val random = SecureRandom()

    /** 32 random bytes as 64 hex characters, and the SHA-256 hex of that string. */
    fun make(): Pair<String, String> {
        val bytes = ByteArray(32).also(random::nextBytes)
        val raw = bytes.toHex()
        return raw to sha256Hex(raw)
    }

    fun sha256Hex(input: String): String =
        MessageDigest.getInstance("SHA-256").digest(input.toByteArray(Charsets.UTF_8)).toHex()

    private fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }
}

/**
 * The exact origin (scheme, host, port) the auth transport trusts, derived from
 * the configured web app URL. HTTPS only, no wildcard, no suffix matching.
 */
class TrustedWebOrigin private constructor(val scheme: String, val host: String, val port: Int) {
    /** The rule for `WebViewCompat.addWebMessageListener`: exactly this origin. */
    val listenerRule: String
        get() = if (port == 443) "$scheme://$host" else "$scheme://$host:$port"

    fun matches(scheme: String?, host: String?, port: Int): Boolean {
        val s = scheme?.lowercase() ?: return false
        val h = host?.lowercase() ?: return false
        val resolvedPort = if (port <= 0) (if (s == "https") 443 else -1) else port
        return s == this.scheme && h == this.host && resolvedPort == this.port
    }

    fun matches(url: String?): Boolean {
        val uri = url?.let { runCatching { URI(it) }.getOrNull() } ?: return false
        return matches(uri.scheme, uri.host, uri.port)
    }

    companion object {
        fun from(url: String): TrustedWebOrigin? {
            val uri = runCatching { URI(url.trim()) }.getOrNull() ?: return null
            val scheme = uri.scheme?.lowercase()
            val host = uri.host?.lowercase()
            if (scheme != "https" || host.isNullOrEmpty()) return null
            return TrustedWebOrigin(scheme, host, if (uri.port <= 0) 443 else uri.port)
        }
    }
}

object GoogleClientConfig {
    private const val SUFFIX = ".apps.googleusercontent.com"

    /**
     * The Web application (server) client ID Google embeds as the ID token
     * audience. Null when missing or malformed, which turns the feature off.
     */
    fun serverClientId(raw: String): String? {
        val value = raw.trim()
        if (!value.endsWith(SUFFIX) || value.length <= SUFFIX.length) return null
        val prefix = value.dropLast(SUFFIX.length)
        return value.takeIf { prefix.all { it.isLetterOrDigit() || it == '-' } }
    }
}

/**
 * Runs callback-style operations strictly one after another in submission
 * order. RevenueCat identify/logout go through it, so a slow logout for the
 * previous account cannot land after the next account's sign-in. Thread-safe.
 */
class SerialCallbackQueue {
    private val lock = Any()
    private val pending = ArrayDeque<(done: () -> Unit) -> Unit>()
    private var running = false

    fun enqueue(operation: (done: () -> Unit) -> Unit) {
        synchronized(lock) {
            pending.addLast(operation)
            if (running) return
            running = true
        }
        runNext()
    }

    private fun runNext() {
        val next = synchronized(lock) {
            pending.pollFirst() ?: run {
                running = false
                null
            }
        } ?: return
        val finished = AtomicBoolean(false)
        next { if (finished.compareAndSet(false, true)) runNext() }
    }
}
